### ISSUE TITLE
Possible KL_loss bug on output dimension average.

### ISSUE THREAD

**Issue by Lan1991Xu** [pytorch/pytorch#6622]
HI I find the KL_loss compute by pytorch is always small than tensorflow. Then I check the defination of KL loss https://en.wikipedia.org/wiki/Kullback%E2%80%93Leibler_divergence, and also the decuments of the http://pytorch.org/docs/master/nn.html, It shows **By default, the losses are averaged for each minibatch over observations as well as over dimensions. However, if the field size_average is set to False, the losses are instead summed.** in pytorch. The problem is the KL loss from wikipedia is not using average dimensions. Any misunderstanding for KL loss implement by pytorch.

**Comment by apaszke** [comment]
KL distance is usually defined as an operator on a pair of distributions, while when you compute a loss you really have multiple distances, and you need to decide how to combine them. PyTorch chooses to average them because this makes gradient magnitudes independent of the batch size, but summation is another valid option. I think our implementation is correct, as you can always switch to the other option.

**Comment by Lan1991Xu** [comment]
Hi apaske, I understand what you means makes gradient magnitudes independent of the batch size. I agree.  But, I  means in this case the KL loss implement by pytorch is also related to **the over dimensions**, for example, if a =[0.1,0.2,0.7],b =[0.2,0.3,0.5], the calculate of KL loss by pytorch need divide by 3 ( which is dimention of B). this seems not consistent with defination of KL distance

**Comment by apaszke** [comment]
Hmmm ok, my bad. I confused that with the negative log likelihood. You have a point that the average might have an incorrect factor, but changing it would be backward breaking (although better mathematically). Any opinions @soumith?

**Comment by Lan1991Xu** [comment]
Actually, I re-implementation KL loss with my side without dividing the dimensions. It works in my case, the really confused me is some other guy using the KL loss (default implement by the Pytorch) it seems work.  But, Actually, I am not understand this  situation. Compared with the orginal version,  if there is two loss(Ep: KL(lossA) , and Cross entropy( loss B). The totall loss  loss A+ lossB. However when you divide the dimensions of KL, it comes to lossA/b +lossB. b is the number of class  in, which will make  large difference in totally loss.

**Comment by apaszke** [comment]
Yes, you're right. We should only average over the batch dimension.

**Comment by Kaixhin** [comment]
Changing the implementation would break a lot of code and the way this has been implemented is consistent with the rest of PyTorch, even if it isn't ideal. Probably best to issue a doc fix? Note that to comply with the standard definition of KL you should use `size_average=False` and then divide by the batch size. Also worth adding explicitly that the target distribution should be standard probabilities (to better contrast that the inputs that are supposed to be log probabilities).

**Comment by t-vi** [comment]
This is #2259 . Personally, I think changing the behaviour, possibly after issuing Warning for a release cycle probably is a good idea, given that averaging probability weights makes very little sense.

**Comment by apaszke** [comment]
Let's add a big warning to the docs for now, but I still think we might want to change this, since as @t-vi noted what we do now doesn't really make sense.

**Comment by ailzhang** [comment]
The doc has already been updated. https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/loss.py#L312
Going to remove the `documentation` label and label it as a bug to fix in 1.0 release.
